### PR TITLE
fix(boot): survive corrupt config & failed migration, surface background errors (#513)

### LINKED ISSUE #513: 1.0.0: boot & runtime resilience (corrupt config, migration guard, renderer error handler)
Found by the pre-1.0.0 multi-agent readiness audit (2026-06-13). The product is solid; these are the genuine stability gaps to close before tagging a *stable* 1.0.0. No new features.

## Tasks

- [ ] **Corrupt/invalid `config.json` bricks boot.** `store` is a module-level `new Store(...)` (`src/main/store.ts:61`) constructed **without** `clearInvalidConfig`. The bundled `conf` engine rethrows on both a JSON `SyntaxError` (corrupt file) and a schema violation when `clearInvalidConfig` is false (default). First store access is in `migrateProfilesToNamedSets()` (`src/main/index.ts:35`) inside `app.whenReady().then(...)` with **no try/catch** -> a corrupted config (power loss mid-write, disk issue, manual edit) means the app will not launch at all: no window, no tray, no dialog. **Fix:** `clearInvalidConfig: true` and/or wrap migrate + store-init in try/catch that renames the bad file, starts fresh, and surfaces a user-facing toast. Add a corrupt-config boot test (none today).

- [ ] **Migration is unguarded, irreversible, and untested against malformed data.** `migrateProfilesToNamedSets()` (`src/main/migrator.ts`) runs unconditionally on every 0.9.x -> 1.0.0 upgrade boot and rewrites the whole `profiles` store in place (`migrator.ts:196`) with no backup. A throw mid-migration propagates out of `whenReady` AND can leave a half-migrated store (`profileSetsMigrated` only set at the end, line 198) -> re-runs next boot against partially-migrated data. **1.0.0 is the upgrade milestone**, so this is the path that matters most. **Fix:** wrap defensively, preserve original `profiles` on failure, add malformed-legacy-data + mid-migration-failure tests (`migrator.test.ts` has zero throw/corrupt cases).

- [ ] **Renderer has no global error / unhandledrejection handler.** The React `ErrorBoundary` only catches render/lifecycle errors. `grep` confirms zero `window.addEventListener('error')` / `onerror` / `unhandledrejection` in `src/renderer`. An async rejection in an event handler or a rejected IPC promise surfaces as an uncaught error in a hidden production window (DevTools closed) -> blank/frozen UI with no Reload. **Fix:** add a window-level error/unhandledrejection listener (log + non-blocking toast, or escalate to the ErrorBoundary fallback).

_Source: pre-1.0.0 readiness audit (completeness critic caught the first two)._

### PR DESCRIPTION
**Pull request by Shieldxx**
First of three PRs from the pre-1.0.0 readiness audit. This is the stability cluster — no new features, all hardening of the frozen feature set.

### 1. Corrupt/invalid `config.json` no longer bricks boot
`store` was a module-level `new Store(...)` with no `clearInvalidConfig`; the bundled `conf` engine **rethrows** on a corrupt JSON or schema violation, and the first access runs in `app.whenReady` with no try/catch → app couldn't launch (no window, no tray, no dialog). New `createResilientStore()` moves the bad file aside (`config.corrupt-<ts>.json`) and retries fresh, falling back to `clearInvalidConfig` if the rename fails. A one-shot notice is pulled by the renderer via a new `get-startup-notice` IPC and toasted, so the user understands why settings reverted to defaults.

### 2. Profile migration can't block startup
`migrateProfilesToNamedSets` is wrapped in try/catch. All store writes happen **after** the migrated shape is fully built, so a throw leaves the original `profiles` untouched and the migrated flags unset — a later launch retries against the original data. (1.0.0 is the upgrade milestone, so this is the path that matters most.)

### 3. Renderer surfaces background errors
The React `ErrorBoundary` only catches render/lifecycle errors. New `window` `error`/`unhandledrejection` handlers (`lib/globalErrors.ts`, installed in the entry point, bridged to the toast system) log every such error and show a toast instead of leaving a blank production window.

## Tests (+8 → 308 total)
- `createResilientStore`: success / quarantine-and-retry / clearInvalidConfig-fallback; `formatConfigRecoveryNotice`
- migrator: a write failure mid-migration doesn't throw, preserves original profiles, leaves flags unset
- `globalErrors`: buffer-before-subscribe + flush, deliver + unsubscribe, unhandledrejection

## Gates
- [x] 308 tests · typecheck · lint · format · build all clean
- [ ] (smoke, at release) confirm the corrupt-config path on a real packaged build

Closes #513.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Closes #513

### CODE REVIEW

**Comment by Shieldxx** [comment]
@codex review

File: src/main/store.ts [hunk 1]
```diff
@@ -84,7 +87,85 @@ export const store = new StoreConstructor({
...
+    const backupPath = quarantine()
+    try {
+      return { store: construct(false), recovery: { backupPath } }
+    } catch {
+      return { store: construct(true), recovery: { backupPath } }
+    }
+  }
+}
+
+const builtStore = createResilientStore(
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/store.ts]: Good catch — fixed in 198f43e. The store is now built lazily on first access via a Proxy, not at import. Every store access runs inside a function that executes after the single-instance lock check (whenReady, IPC handlers), and a second instance quits before any access, so the corrupt-config recovery only ever runs for the primary instance. Verified no module-level store access at import; added an isolated test asserting import doesn't construct the store.

**Comment by Shieldxx** [comment]
@codex review

File: src/main/store.ts [hunk 2]
```diff
@@ -84,7 +87,109 @@ export const store = new StoreConstructor({
...
+ * recovery — which can rewrite/quarantine config.json — only ever runs for the
+ * PRIMARY instance. index.ts imports `store` before requestSingleInstanceLock();
+ * a second launch quits on the lock before any store access, so it never builds
+ * the store and can't touch the live user's config (#516 Codex P2). Every store
+ * access in the app runs inside a function that executes after the lock check.
+ */
+export const store = new Proxy({} as StoreInstance, {
+  get(_target, prop, receiver) {
+    const instance = ensureStore()
+    const value = Reflect.get(instance as object, prop, receiver)
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/store.ts]: Good catch — fixed in c96f4ad. The Proxy now forwards reads with the real `instance` as the `Reflect.get` receiver instead of the Proxy, so electron-store's `.store`/`.path`/`.size` getters (which read private `#` fields) resolve against a genuine `Conf` instance. Verified the two real call sites — the import-rollback snapshot (`{ ...store.store }`) and config export (`getSupportedConfigValues(store.store)`) — and added a regression test in `tests/main/storeLazy.test.ts` using a mock whose `store` getter reads a private field, which throws if the receiver is the Proxy.

**Comment by Shieldxx** [comment]
@codex review

File: src/main/ipc/config.ts [hunk 3]
```diff
@@ -9,6 +9,8 @@ import {
   KNOWN_GAME_KEYS,
   MAX_CONFIG_IMPORT_BYTES,
   MAX_CUSTOM_SLOTS,
+  consumeConfigRecoveryNotice,
+  formatConfigRecoveryNotice,
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/ipc/config.ts:13]: Hardened in 3ff3e3c. One correction on the failure mode: these suites do still run and pass today (the latest CI run and a local run both show `configImport`/`configImportPreview`/`configZoom` green) — with `vi.doMock` the missing named exports resolve to `undefined` rather than failing module load, and none of these suites exercise the `get-startup-notice` handler that actually calls them, so nothing is invoked. That said, the point stands that the mocks no longer mirror the real export surface, which is latent fragility, so I've added `consumeConfigRecoveryNotice` and `formatConfigRecoveryNotice` to all three store mocks.

**Comment by Shieldxx** [comment]
@codex review

File: src/main/migrator.ts [hunk 4]
```diff
@@ -154,6 +154,18 @@ function normalizeStoredProfileSet(profileEntry: StoredProfileEntry, utilityKeys
 }
 
 export function migrateProfilesToNamedSets(): void {
+  try {
+    runProfileSetMigration()
+  } catch (error) {
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/migrator.ts]: Real bug — fixed in 1c8a674. You're right that `applySanitizedConfig` depends on the migration throwing to restore its pre-import snapshot, and the blanket try/catch I added broke that: a post-migration `store.set` failure during import would have reported success over a half-written store.

Rather than special-case import, I moved the resilience to where it belongs. `migrateProfilesToNamedSets` now propagates failures again (its natural contract); config import keeps its throw-based rollback unchanged; and the boot call site in `index.ts` wraps the call in its own try/catch so a malformed legacy profile still can't brick startup (every store write lands only after the migrated shape is fully built, so a throw leaves the original profiles and the migrated flags untouched and a future launch retries).

Updated the migrator test to assert the throw and added a boot test proving startup survives a throwing migration.

**Comment by Shieldxx** [comment]
@codex review

**Comment by Shieldxx** [comment]
## Adversarial self-audit before merge

Given this PR touches the boot/config recovery seam and earlier review rounds each surfaced a real bug, I ran an independent adversarial audit of the error paths (7 review lenses — lazy store Proxy, corrupt-config recovery/quarantine, migration throw/rollback contract, import rollback snapshot, renderer global errors + StrictMode, startup-notice IPC, and test-coverage gaps — each finding then verified by 3 refute-by-default skeptics).

It found **one** confirmed defect (3/3 skeptics), fixed in 70204e7:

**P2 — a locked/permission-denied `config.json` still bricked boot.** `createResilientStore` wrapped the first two construction attempts in try/catch but let the final `construct(true)` throw, assuming `clearInvalidConfig: true` always resets in place. It doesn't: `conf` only swallows `SyntaxError`/schema/decrypt errors, so a locked or permission-denied file (EBUSY/EPERM from an AV scan, a sync client, or a second instance mid-write) rethrows from the constructor and escapes `ensureStore()` uncaught. The boot migration try/catch then swallowed and mislabeled the first throw, and `store.get('showTrayIcon')` re-threw inside `whenReady().then()` with no `.catch` → no window, no tray, no dialog.

Fix: a fourth recovery rung boots on an ephemeral in-memory store seeded with schema defaults when even `clearInvalidConfig` throws, so `ensureStore()` never throws (which also removes the mislabeling and the unhandled rejection). The startup notice gains an `ephemeral` variant that, unlike the corrupt-config reset, correctly tells the user nothing was changed and their settings reload next launch. +2 tests (ephemeral ladder rung + in-memory fallback behavior); 313 green.

The other 4 candidate findings were all "missing test" suggestions that the skeptics judged non-defects (≤1/3) — noted as optional follow-ups, not blockers.

@codex review

File: src/main/store.ts [hunk 5]
```diff
@@ -84,7 +87,189 @@ export const store = new StoreConstructor({
...
+// uses (get/set/clear/`store`) is implemented, and the methods read closure state
+// rather than `this`, so the lazy Proxy can bind/forward them safely.
+export function createInMemoryFallbackStore(): StoreInstance {
+  const schema = STORE_OPTIONS.schema as Record<string, { default?: unknown }> | undefined
+
+  const seedDefaults = () => {
+    const seeded: Record<string, unknown> = {}
+    Object.entries(schema ?? {}).forEach(([key, spec]) => {
+      if (spec && 'default' in spec) {
+        seeded[key] = spec.default
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/store.ts]: Good catch — fixed in 3833d04. `seedDefaults()` now `structuredClone`s each default, so the fallback never hands out or reseeds from the shared `STORE_OPTIONS.schema` reference. Confirmed the aliasing was real: `save-profile` does `profiles[gameKey] = ...; store.set('profiles', profiles)` on the object from `store.get('profiles')`, which in the fallback was the schema default. Extended the fallback test to mutate a returned object default, `clear()`, and assert both this instance and a fresh one reset to `{}`.

**Comment by Shieldxx** [comment]
@codex review
